guardian/cdk
# feat(riff-raff.yaml): Support multiple image recipes

## What does this change?


Supports the use-case where there are multiple ASGs in a single CloudFormation template, each using a different image recipe. This requires moving from `amiParameter` and `amiTags` to `amiParametersToTags` ([docs](https://public.riffraff.gutools.co.uk/docs/magenta-lib/types#cloudformation)). See the added test for a complete example.

~Also add support for encrypted AMIs, via a new `imageRecipeEncrypted` property on both `GuEc2App`, and `GuAutoscalingGroup`.~ Update: These changes have been [reverted](https://github.com/guardian/cdk/pull/1697/commits/1333a22e3bcc96c3abc68d61c8c7196feb37ba20) as it appears [Riff-Raff doesn't follow](https://github.com/guardian/riff-raff/blob/3f580f70795c7b210e4085cb8d16fa95cfcd1321/magenta-lib/src/main/scala/magenta/deployment_type/CloudFormationDeploymentTypeParameters.scala#L161-L164) the `Encrypted` tag when set in `amiParametersToTags`. Recommend we add this change in a separate PR.

cc @twrichards, @dskamiotis who spotted this when developing the new fingerpost service - https://github.com/guardian/editorial-wires/pull/13.

## How to test



See added test, and validate the generated YAML (see below) [in Riff-Raff](https://public.riffraff.gutools.co.uk/configuration/validation). 

See also https://github.com/guardian/cdk-playground/pull/325 for a test on a real-world application.

<details>
<summary>Generated YAML</summary>

```yaml
allowedStages:
  - CODE
deployments:
  asg-upload-eu-west-1-test-my-api:
    type: autoscaling
    actions:
      - uploadArtifacts
    regions:
      - eu-west-1
    stacks:
      - test
    app: my-api
    parameters:
      bucketSsmLookup: true
      prefixApp: true
    contentDirectory: my-api
  asg-upload-eu-west-1-test-my-data-collector:
    type: autoscaling
    actions:
      - uploadArtifacts
    regions:
      - eu-west-1
    stacks:
      - test
    app: my-data-collector
    parameters:
      bucketSsmLookup: true
      prefixApp: true
    contentDirectory: my-data-collector
  cfn-eu-west-1-test-my-application-stack:
    type: cloud-formation
    regions:
      - eu-west-1
    stacks:
      - test
    app: my-application-stack
    contentDirectory: /tmp/cdk.out
    parameters:
      templateStagePaths:
        CODE: test-stack.template.json
      amiParametersToTags:
        AMIMyapi:
          BuiltBy: amigo
          Recipe: arm64-bionic-java11-deploy-infrastructure
          AmigoStage: PROD
          Encrypted: false
        AMIMydatacollector:
          BuiltBy: amigo
          Recipe: arm64-bionic-node18-deploy-infrastructure
          AmigoStage: PROD
          Encrypted: true
    dependencies:
      - asg-upload-eu-west-1-test-my-api
      - asg-upload-eu-west-1-test-my-data-collector
  asg-update-eu-west-1-test-my-api:
    type: autoscaling
    actions:
      - deploy
    regions:
      - eu-west-1
    stacks:
      - test
    app: my-api
    parameters:
      bucketSsmLookup: true
      prefixApp: true
    dependencies:
      - cfn-eu-west-1-test-my-application-stack
    contentDirectory: my-api
  asg-update-eu-west-1-test-my-data-collector:
    type: autoscaling
    actions:
      - deploy
    regions:
      - eu-west-1
    stacks:
      - test
    app: my-data-collector
    parameters:
      bucketSsmLookup: true
      prefixApp: true
    dependencies:
      - cfn-eu-west-1-test-my-application-stack
    contentDirectory: my-data-collector
```
</details>

## How can we measure success?


Supporting more use-cases.

## Have we considered potential risks?


N/A. This change is backwards compatible with previous versions.

## Checklist


[^1]: Consider whether this is something that will mean changes to projects that have already been migrated, or to the CDK CLI tool. If changes are required, consider adding a checklist here and/or linking to related PRs.
[^2]: If you are adding a new construct or pattern, has new documentation been added? If you are amending defaults or changing behaviour, are the existing docs still valid?
